Sources/XCTest/Public/XCTestCase.swift

```diff
@@ -136,13 +153,140 @@ open class XCTestCase: XCTest {
             }
         }
 
+        do {
+            if skip == nil {
+                try unwrapAndInvokeTestClosure()
+            }
+        } catch {
+            handleError(error)
+        }
+
         if let skip = skip {
             testRun?.recordSkip(description: skip.summary, sourceLocation: skip.sourceLocation)
         }
+    }
+
+    private func unwrapAndInvokeTestClosure() throws {
+        let expectation = XCTestExpectation(description: "invokeTest completion")
+
+        @Sendable func handleError(_ error: Error) {
+            
+            if error.xct_shouldSkipTestInvocation && self.skip == nil {
+                if let skip = error as? XCTSkip {
+                    self.skip = skip
+                } else {
+                    self.skip = XCTSkip(error: error, message: nil, sourceLocation: nil)
+                }
+            }
+
+            if error.xct_shouldRecordAsTestFailure {
+                self.recordFailure(for: error)
+            }
+        }
+        
+        switch testClosure {
+        case .regular(let closure):
+            
+            do {
+                try setUpWithError()
+            } catch {
+                handleError(error)
+                return
+            }
+            
+            setUp()
+            
+            try closure(self)
+            
+            do {
+                try tearDownWithError()
+            } catch {
+                handleError(error)
+                return
+            }
+            
+            tearDown()
+            
+            
+            
+            Task {
+                defer { expectation.fulfill() }
+                await runTeardownBlocks(errorHandler: handleError(_:))
+            }
+            
+        case .async(let closure):
+            
+            Task {
+                defer { expectation.fulfill() }
+                
+                try await withThrowingTaskGroup(of: Void.self) { group in
+                    group.addTask {
+                        do {
+                            
+                            do {
+                                try await self.setUp()
+                                try self.setUpWithError()
+                            }
+                            
+                            
+                            let nonAsyncSetUp: () -> Void = self.setUp
+                            nonAsyncSetUp()
+                            
+                            try await closure(self)
+                            
+                        } catch {
+                            handleError(error)
+                        }
+                    }
+                    
+                    group.addTask {
```

Comment:
I don't think we can safely use the `withThrowingTaskGroup` or `group.addTask` APIs here since it means their order of execution is unspecified, so for example tearDown could end up running before or concurrently with `setUp`. I think we'll probably want to have a code flow similar to that of the `.regular` closure type above

```swift
  /// Unconditionally create a child task in the group.
  /// 
  /// The child task will be executing concurrently with the group, and its result 
  /// may be collected by calling `group.next()` or iterating over the group gathering 
  /// all submitted task results from the group.
  mutating func addTask(
```
https://github.com/apple/swift-evolution/blob/main/proposals/0304-structured-concurrency.md#creating-taskgroup-child-tasks